### DISCUSSION TITLE
ERROR: Key is not a valid public or private key

### DISCUSSION THREAD

**Discussion by favoooo** [DigitalAXPP/jwtPS#41] [Q&A] [answered]
Hello,

I am receiving the error: Key is not a valid public or private key. 

The private key has worked with a javascript library I was testing.

What could I possibly be doing wrong? Any help is much appreciated.

```
$key = "<PATH>";

$keyContent = (Get-Content -Path $key) -join ""

$header = @{
    alg = "ES256"
    typ = "JWT"
    kid = "<XXX>"
}


$payload = @{
    aud = "<XXXX>"        
    iss = "<XXXX>"
    scope = "rest_webservices"            
    exp = ([System.DateTimeOffset]::Now.AddHours(3)).ToUnixTimeSeconds()
    iat = ([System.DateTimeOffset]::Now).ToUnixTimeSeconds()
}
$encryption = [jwtTypes+encryption]::SHA256
$algorithm = [jwtTypes+algorithm]::ECDsa
$alg = [jwtTypes+cryptographyType]::new($algorithm, $encryption)
$jwt = New-Jwt -Payload $payload -Algorithm $alg -Header $header -Secret $keyContent
```

**Comment by DigitalAXPP** [comment]
Hi Favian,
I ran your code and I was able to generate a JWT with the test ECDsa .pem file.
`eyJhbGciOiJFUzI1NiIsImtpZCI6IiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDQ4MzI4MjMsImlhdCI6MTcwNDgyMjAyMywiaXNzIjoiIiwiYXVkIjoiIiwic2NvcGUiOiJyZXN0X3dlYnNlcnZpY2VzIn0.hSPoeQsqImcIX5OQO_e7O0K-nWwIL0IHSUreOL3iVVqFml32EX10UxPP061npOdl_pRGOg382m28aBPMAz3bkA`

Can you check whether your key has the same structure as my key (.\github\workflows\private_es256.pem)?

**Reply by favoooo** [reply]
Hello,

Thank you for getting back to me. To answer your question, somewhat, my key is quite larger. I have been able to generate a JWT token since I posted. However, jwt.io is unable to validate the signature. I am struggling to figure out where I am going wrong. I am also just starting to get familiar with the concept so I'm sure that doesn't help.

Here is the command I used to generate the certificate and key:

`openssl req -new -x509 -newkey rsa:4096 -keyout private.pem -sigopt rsa_padding_mode:pss -sha256 -sigopt rsa_pss_saltlen:64 -out public.pem -nodes`

**Reply by DigitalAXPP** [reply]
Hi Favoo,
I created my own key now and used PSS as algorithm which allowed me to create the jwt.
In jwt.io the signature was however also invalid. I assume you tried PSS as well.

Maybe the public key format is not correct and that's why the JWT cannot be validated in jwt.io.

Can you try to authenticate to your website to check whether it works?

(P.S I am not an expert on certificates and keys either.)

**Reply by favoooo** [reply, marked as answer]
Hello DigitalAXPP,

I was actually able to get this working by tweaking my code a bit. In the '$header' payload I change the 'alg' to 'PS256', and also changed the '$algorithm' to 'PSS'. The token signature generated using this combination was successfully verified by jwt.io, and allowed me to retrieve the access token needed for my application.

```
$key = "<PATH>";

$keyContent = (Get-Content -Path $key) -join ""

$header = @{
    alg = "PS256"
    typ = "JWT"
    kid = "<XXX>"
}


$payload = @{
    aud = "<XXXX>"        
    iss = "<XXXX>"
    scope = "rest_webservices"            
    exp = ([System.DateTimeOffset]::Now.AddHours(1)).ToUnixTimeSeconds()
    iat = ([System.DateTimeOffset]::Now).ToUnixTimeSeconds()
}
$encryption = [jwtTypes+encryption]::SHA256
$algorithm = [jwtTypes+algorithm]::PSS
$alg = [jwtTypes+cryptographyType]::new($algorithm, $encryption)
$jwt = New-Jwt -Payload $payload -Algorithm $alg -Header $header -Secret $keyContent
```

Thanks for your help!

**Reply by DigitalAXPP** [reply]
Hi Favoo,
That's right and I even wrote this down in the Wiki section to consider when one creates a custom header.
New improvement idea: warning message when `alg` doesn't match the algorithm!

Thanks a lot for your feedback!
